### ISSUE TITLE
QNN-delegated static-cache decoder produces correct output via native C++ Module, but degenerate/repeating output via Android JNI Module (same .pte, same device)

### ISSUE THREAD

**Issue by aramvyan** [pytorch/executorch#22061]
# QNN-delegated static-cache decoder produces degenerate output through Android JNI `Module`, but correct output through the native C++ `Module`/runner — same `.pte`, same device

## Summary

A Whisper `.pte` exported with `to_edge_transform_and_lower_to_qnn()` (QNN HTP backend, static KV-cache decoder) produces correct, real transcripts when run through ExecuTorch's native C++ `Module` API (`examples/qualcomm/oss_scripts/whisper/runner/`), but produces degenerate, repetition-loop garbage when run through the Android JNI `Module`/`Tensor`/`EValue` wrapper (`extension/android/`) — same `.pte` file (verified byte-identical via sha256), same physical device, same QNN SDK version, same greedy-decode algorithm (verified line-by-line equivalent).

## Environment

- ExecuTorch: checked out fresh (commit at time of testing — pin to whatever was `HEAD` on the build box; not recorded precisely, should be captured before filing)
- QNN SDK: 2.37.0.250724
- Device: nubia NX789J-UN (Snapdragon 8 Elite Gen5, SM8750), Android 15
- HTP: v79 (`libQnnHtpV79Skel.so`/`libQnnHtpV79Stub.so`)
- Model: Whisper tiny, exported via `examples/qualcomm/oss_scripts/whisper/whisper.py` (16a8w quantization, PT2E, `MinMaxObserver`, per-channel conv+linear)

## What works: native C++ path

Built `qnn_whisper_runner` via the official `backends/qualcomm/scripts/build.sh --skip_x86_64`, pushed to the device with the matching QNN runtime `.so`s (`libQnnHtp.so`, `libQnnHtpV79Stub.so`, `libQnnHtpPrepare.so`, `libQnnSystem.so`, `libQnnHtpV79Skel.so`), ran directly via `adb shell` (bare `shell` UID):

```
./qnn_whisper_runner --model_path whisper_tiny_qnn_16a8w.pte --tokenizer_json_path tokenizer.json \
  --input_list_path input_list.txt --seq_len 32 --output_folder_path outputs
```

`input_0_0.raw` was a `[1,80,3000]` float32 mel tensor computed via `transformers.AutoProcessor` on an 11s English speech clip (JFK inaugural excerpt).

**Result** (`outputs/output_0.txt`):
```
<|en|><|transcribe|><|notimestamps|> And so my fellow Americans ask not what your country can do
for you ask what you can do for your country.<|endoftext|>
```
27 tokens, real varying vocabulary throughout, correct structural token at position 3
(`<|notimestamps|>`, id 50363), clean EOS termination. Model load 0.44s, encode 0.032s, decode
0.159s — confirmed real HTP execution via `[INFO] [Qnn ExecuTorch]` logs, not a CPU fallback.

(Note: the two forced-token positions, `<|en|>`/`<|transcribe|>` at decoder positions 1/2, are an
application-level convention from the consuming Android app, replicated in the runner via a small
patch to `runner.cpp` for apples-to-apples comparison with the Android path below — not upstream
behavior. The important part is what happens from position 3 onward, which is entirely the model's
own unforced greedy output either way.)

## What's broken: Android JNI path

Same `.pte` (verified byte-identical, sha256 match), loaded via `org.pytorch.executorch.Module` from
Kotlin:

```kotlin
val module = Module.load(modelFile.absolutePath, Module.LOAD_MODE_FILE)
module.loadMethod("encoder")
module.loadMethod("decoder")
// ... greedy decode loop, one execute() call per token:
val logits = module.execute(
    "decoder",
    EValue.from(Tensor.fromBlob(longArrayOf(curToken), longArrayOf(1, 1))),
    EValue.from(Tensor.fromBlob(attentionMask.copyOf(), longArrayOf(1, 1, 1, maxContextLen))),
    encoderOutValue,
    EValue.from(Tensor.fromBlob(longArrayOf(pos), longArrayOf(1))),
)[0].toTensor().dataAsFloatArray
```

Algorithm (attention-mask indexing, `pos` tracking, argmax selection) verified line-by-line
equivalent to the C++ runner's loop. Same device, same `.pte`, run moments apart.

**Result, every run, all 6 Whisper sizes tested (tiny/base/small/medium/large-v3/large-v3-turbo)**:
degenerate output starting at decoder position 3 (the first position not forced by the calling
app), typically a repetition loop of a single token within 2-5 total generated tokens. Examples
observed on `tiny` across many runs on the same fixture: `<|startoftranscript|>` repeating,
`<|sv|>` repeating, `<|transcribe|>` repeating (varies run to run, and varied with different build
configurations — see below — but never produces real content). One data point of note:
`large_v3`'s first free prediction was the *correct* structural token, `<|notimestamps|>` — then it
got stuck repeating that same token instead of advancing, rather than producing an implausible
token outright.

## running on real hardware, not just reasoned about)

1. **Calibration data quality/representativeness** — an A/B between the model's original
   (unpatched, fully free-running) calibration and a corrected, properly-forced-token calibration
   produced byte-for-byte identical failure. Calibration data is not the cause.
2. **Calibration missing a repetition-early-stop** (unlike the consuming app's own decode loop,
   calibration only breaks on EOS) — patched calibration to add the same repetition-stop, re-ran
   full export + on-device test. No change.
3. **`skip_mutable_buffer`** (`to_edge_transform_and_lower_to_qnn()`'s QNN-partitioner flag,
   documented for exactly "accuracy issue during the runtime") — tested `True` vs default `False`.
   No change.
4. **`Module` load mode** — tested `LOAD_MODE_MMAP_USE_MLOCK_IGNORE_ERRORS` (matching the native
   runner's `Module::LoadMode::MmapUseMlockIgnoreErrors` exactly) vs the app's original
   `LOAD_MODE_FILE`. No change.
5. **Single shared `Module` for encoder+decoder vs two independent `Module` instances** (matching
   the native runner's `WhisperEncoder`/`WhisperDecoder` two-object split exactly, each loading the
   same `.pte` independently). No change.
6. **On-device mel-spectrogram computation** (the app's own `whisper.cpp`-based
   `whisper_pcm_to_mel`) vs the Python `AutoProcessor`-computed features used in the native-runner
   test — dumped the app's real computed mel tensor and compared numerically against the Python
   one: max abs diff 1.88e-5, correlation 0.99999999999981 (float32 noise only). Not the cause.
7. **QNN runtime `.so` version mismatch** — the app's bundled `libQnnHtp.so`/
   `libQnnHtpPrepare.so`/`libQnnHtpV79Skel.so`/`libQnnHtpV79Stub.so` are sha256-identical to the
   ones used in the successful native-runner test. Not the cause.
8. **A stale/outdated `executorch-android` AAR** — rebuilt the AAR from scratch via the official
   `scripts/build_android_library.sh` against the same checkout used for the successful native
   runner (installed a full Android SDK inside the same build environment for this). The freshly
   built `libqnn_executorch_backend.so` came out byte-identical to the currently-vendored one
   (build-configuration-deterministic, not stale). `libexecutorch.so` *did* differ from the
   currently-vendored one. Swapped the full freshly-built matched pair into the app, rebuilt,
   reinstalled, tested: **behavior measurably changed** (different specific repeating token, 4
   generated tokens instead of 5) but the core symptom — immediate degenerate output starting right
   after the two forced positions — persisted.
9. **Zero-copy JNI tensor lifetime** — `extension/android/jni/jni_layer.cpp`'s
   `TensorHybrid::newJTensorFromTensor()` wraps the *output* tensor's `data_ptr()` directly in a
   zero-copy Java `ByteBuffer` (`facebook::jni::JByteBuffer::wrapBytes`), with an explicit code
   comment: safe only "as long as ... there is no other inference call" before the Java side reads
   it. The app's decode loop holds and reuses the `encoder` method's output `Tensor` across every
   subsequent `decoder` `execute()` call in the loop (i.e., many "other inference calls" after the
   one that produced it) — exactly the scenario the comment warns about. Deep-copied the encoder
   output into an independently-owned Kotlin `FloatArray`/`Tensor` before the decode loop instead of
   holding the zero-copy view. No change.

Also checked and ruled out: cross-thread execution (the consuming app runs model-load through every
encode/decode call on one dedicated single-thread `Executor` — never multi-threaded).

## What we could not cleanly test

Tried running the native `qnn_whisper_runner` binary under the consuming app's own UID/SELinux
domain (`adb shell run-as <pkg> ...`, since `/data/local/tmp` binaries can't execute under
`untrusted_app` SELinux) to see whether HTP/FastRPC transport behaves differently there than under
bare `shell`. This hit a *different* failure (`QnnDsp <E> Failed to load skel, error: 4000` /
`Failed to create transport for device, error: 4000`) regardless of which native-library directory
was used (including the app's real, properly-installed `nativeLibraryDir`) — but the real,
production Android app *does* get genuine HTP delegate access every single run (confirmed via
`MethodMetadata.getBackends()` reporting `QnnBackend`, and real NPU-speed timings, never a CPU
fallback signature), so this specific `run-as`-executed-binary test isn't a valid apples-to-apples
comparison — it's very likely missing some Android-lifecycle-level FastRPC/vendor-binder
registration step that only happens through normal `Context`/`Application` initialization, not
through a bare `run-as exec`. Abandoned this angle as inconclusive rather than chase a second,
apparently-unrelated failure mode.

## Ask

Is there a known difference between how the native `Module`/backend delegate handles are
initialized/persisted versus how `extension/android`'s JNI wrapper does, specifically for models
using QNN-delegated **mutable buffers** (static KV-cache decoders)? The QNN partitioner's own
`skip_mutable_buffer=False` (default) code comment documents a "same I/O address" optimization for
delegated mutable buffers during the runtime — is there anything about crossing the JNI boundary
(tensor construction via `facebook::jni::JByteBuffer`, `Module` object lifetime under ART's GC,
thread-affinity assumptions) that could break that address-stability assumption in a way a bare C++
call stack wouldn't?

Happy to provide the exact `.pte`, a minimal reproduction Kotlin/JNI harness, or run additional
diagnostics on request — this has been isolated fairly thoroughly on the calling-app side (9
independent hypotheses tested by actually rebuilding and running on real hardware, not just static
analysis) and increasingly looks like it needs an answer from inside ExecuTorch's own JNI/QNN
backend internals rather than anything fixable from the Android app or export-script side.

cc @kirklandsign @cbilgin @psiddh

**Comment by shewu-quic** [comment]
Hi @aramvyan, 
Thanks for the detailed analysis and for taking the time to investigate this issue. Based on your findings, it doesn't seem to be a problem in the QNN backend itself, since the example script produces the correct output.

As a next step, would it be possible to further narrow down where the accuracy issue is introduced with jni interface? For example, can we determine whether it already exists in the encoder output, or if it only appears in the decoder path?

I'm asking because the decoder model is the only part that uses skip_mutable_buffer, so understanding whether the issue is encoder-specific or decoder-specific could help further investigation.

**Comment by aramvyan** [comment]
Thanks @shewu-quic — direct answer, plus several more hypotheses tested since filing, all with the
same real-hardware-rebuild-and-measure rigor as the original 9. Short version: it's decoder-specific,
and it's **not** about mutable-buffer address stability specifically — a redesign with zero mutable
buffers reproduces the identical symptom.

## Encoder vs decoder: confirmed encoder-only, deterministic

Ran the `encoder` method alone, 50x back-to-back on the real device (same mel input every time, no
decoder involved at all), hashing the raw output tensor bytes each run:

```
RESULT verdict=BIT_IDENTICAL runs=50 uniqueHashes=1
referenceSha256=2021fb24b1727236eff71d58c45d7c29238714e27a1b41814957f6a4078ba220
shape=[1, 1500, 384]
```

50/50 bit-identical. The encoder is fully deterministic through the same Android `Module` API that
produces garbage from the decoder. Whatever's wrong is entirely inside the decoder method's own
execution/state, not encoder output corruption propagating downstream.

## New architecture tested: single JNI crossing (rules out hypothesis #9)

Since filing, we replaced the whole per-token JNI loop (`Module`/`Tensor`/`EValue`, one JNI crossing
per generated token) with a custom native entry point that does the *entire* encode+decode greedy
loop inside one JNI call — Kotlin passes in a mel tensor once and gets back the full token sequence,
no `Tensor`/`EValue` objects cross the JNI boundary more than once, no possibility of the zero-copy
output-buffer lifetime issue hypothesis #9 describes (there's no "another inference call" happening
after a JNI-visible `Tensor` is handed to Java — the whole loop is C++-only).

**Identical failure signature.** Same repetition-loop garbage, same distinctive pattern (below).
Hypothesis #9's mechanism can be ruled out entirely — this isn't about anything crossing the JNI
boundary repeatedly, since it no longer does.

## New architecture tested: fully external KV-cache, zero mutable buffers

This one directly targets your `skip_mutable_buffer` question. We re-exported the decoder with the
self-attention KV-cache made **fully external** — 8 explicit per-layer input/output tensors
(`k_cache_self_N`/`v_cache_self_N`) instead of QNN-delegated mutable buffers at all, with the
*application* doing an explicit host-side `memcpy` of each step's cache outputs into the next step's
cache inputs (zero-initialized once per session) — i.e., no `skip_mutable_buffer=False`
address-stability mechanism in play whatsoever, because there are no mutable buffers in this export.

(Getting this to lower to QNN required two op-level fixes worth noting for anyone hitting the same
thing: `index_copy` for the cache write gets decomposed to a general `index_put`/ScatterNd path that
isn't QNN-delegatable — swapping to `torch.scatter` fixes that. And a plain `float * bool` multiply
used to build the attention mask from `cache_position` — needed because cache position is now a real
external input rather than derived from hidden state — also isn't delegatable, and an explicit
`.to(dtype)` cast on the bool operand *also* isn't; `torch.where(bool_cond, a, b)` is what QNN
actually accepts for this pattern. Neither of those two things affected the actual bug either way,
just export mechanics.)

**Identical failure signature again.** Zero mutable buffers, fully explicit application-managed
state round-tripped through ordinary I/O tensors every step, same garbage.

## The failure signature itself, characterized further

Every failing run stops via a repetition-loop guard within 3-6 generated tokens. The app forces
exactly one decoder position (position 2, the task token) via post-argmax override; that position is
always correct 100% of the time, success or failure — expected, since it's not actually computed.
Position 1 (language token, otherwise free/unconstrained) is where success/failure diverges: on
success it's a real predicted language token; on failure it's consistently the *decoder's own start
token id* (i.e., `argmax` at the very first free position lands on the same token id that was fed in
as `decoder_start_token_id`).

We tested whether this is the model "echoing" whatever token id is fed as input: swapped
`decoder_start_token_id` to a different valid token id (the EOS id) for one test run. If it were
echoing the input, we'd expect the garbage to now land on the new value. It didn't — 10/10 runs still
produced the *original* start-token id at position 1, unchanged despite the different input. So it
isn't reading back its own input; something about the first free position's logits computation
specifically produces a fixed, input-independent wrong answer.

## Also tried, no change either way

- Explicit `rpcmem`/QNN-shared-memory registration for every decoder I/O tensor + logits output
  (`QnnExecuTorchAllocCustomMem`/`QnnExecuTorchAddCustomMemTensorAddr`, proper FD/memhandle path, not
  a naive allocator swap) on the mutable-buffer version — no change.
- Zero-initializing the mutable-buffer allocation at the ExecuTorch/QNN backend level
  (`QnnManager`/`TensorWrapper`'s `new char[bytes]` → `new char[bytes]()`) — no change.
- Forcing the language token too (matching how the task token is already forced) — sidesteps the
  corrupted read at that specific position but doesn't fix anything upstream; no change to overall
  reliability.

## Rough numbers

Across several measurement runs on this device (same fixture, `tiny`, English), correct-transcript
rate has ranged roughly 0–25% depending on exact build/config, never approaching consistent success,
and every failing run shows the same position-1 signature described above regardless of which of the
above variants was under test.

## Updated ask

Given a fully mutable-buffer-free export shows the identical symptom, is there a different shared
mechanism between the native `Module`/runner path and the Android JNI `Module` path that could
produce a deterministic-per-invocation-but-wrong-across-invocations result specifically at the first
unconstrained decode position — independent of whether KV-cache is QNN-internal or fully external?
Something in how the Android JNI `Module`'s underlying QNN context/session gets initialized or
first-executed differently than the native runner's, that would affect exactly the first free logits
computation and nothing else (later positions, when reached, are also usually corrupted once the
sequence has gone off the rails, but the *first* divergence is always right here)?

Happy to share the exact instrumented `.pte`/build, or run further diagnostics you'd want to see —
this has now survived a fairly wide range of structural changes to the decode path (JNI crossing
count, mutable vs. fully-external cache, shared-memory registration) without changing, which suggests
whatever's wrong sits somewhere more fundamental to Android JNI `Module` + QNN HTP session
init/execution than any of the app-side mechanisms tested so far.

**Comment by psiddh** [comment]
Thanks for the very detailed investigation. 

Could you share the exact ExecuTorch commit, .pte, tokenizer, and the mel input (input_0_0.raw) ? I am thinking if we  can build the minimal Android  test harness and use these artifacts to compare the encoder output and first decoder step across both paths. Cross-feeding each path’s encoder output into the other decoder should identify whether the divergence comes from the model inputs, native build, or Android process environment. What do you think ?

**Comment by ashish-frozo** [comment]
@aramvyan this is a strong repro because it already separates “QNN works” from “the Android app path preserves behavior.”

The details I’d anchor on are:

- same `.pte`, same physical device, same QNN SDK `2.37.0.250724`
- native C++ QNN path works on HTP: model load `0.44s`, encode `0.032s`, decode `0.159s`
- Android JNI path goes into repetition-loop garbage across all 6 Whisper sizes
- encoder-only through Android `Module` is stable: `RESULT verdict=BIT_IDENTICAL runs=50 uniqueHashes=1`
- the failure is decoder-specific, and the mutable-buffer-free export still reproduces it

I agree with @psiddh’s cross-feed direction. The thing I’d add is making the comparison device-evidence-backed: fixed mel input, native encoder -> Android decoder, Android encoder -> native decoder, first decoder logits, and actual HTP/CPU residency attached to each run. Otherwise you can prove the artifact is identical and still miss whether the runtime path changed under the app process.

This is the exact class of check we’re building EdgeGate around: fail CI when device-backed QNN behavior or NPU residency changes before it gets into the app path. Want a quick NPU residency check on that model path? https://edgegate.frozo.ai

**Comment by nil-is-all** [comment]
> Thanks for the very detailed investigation.
> 
> Could you share the exact ExecuTorch commit, .pte, tokenizer, and the mel input (input_0_0.raw) ? I am thinking if we can build the minimal Android test harness and use these artifacts to compare the encoder output and first decoder step across both paths. Cross-feeding each path’s encoder output into the other decoder should identify whether the divergence comes from the model inputs, native build, or Android process environment. What do you think ?

Hi @aramvyan, could you share the commit and other info to move forward?
